ssh-server/main.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"bytes"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+
+	"github.com/kr/pty"
+	"golang.org/x/crypto/ssh"
+)
+
+var (
+	serverPort = flag.Uint("port", uint(22), "the port to listen on")
+)
+
+func main() {
+	flag.Parse()
+
+	config, listener, err := startServer(*serverPort)
+	if err != nil {
+		log.Fatalf("Failed to start server: %v", err)
+	}
+	log.Printf("Listening on %s...", listener.Addr())
+
+	for {
+		connection, err := listener.Accept()
+		if err != nil {
+			log.Fatalf("Failed to accept incoming connection: %v", err)
+		}
+
+		go func() {
+			handleConnection(connection, config)
+		}()
+	}
+}
+
+func startServer(port uint) (*ssh.ServerConfig, net.Listener, error) {
+	config, err := getConfiguration()
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting configuration: %v", err)
+	}
+
+	serverAddress := fmt.Sprintf(":%d", port)
+	listener, err := net.Listen("tcp", serverAddress)
+	if err != nil {
+		return nil, nil, fmt.Errorf("listen: %v", err)
+	}
+	return config, listener, nil
+}
+
+func getConfiguration() (*ssh.ServerConfig, error) {
+	config := &ssh.ServerConfig{NoClientAuth: true}
+
+	privateBytes, err := ioutil.ReadFile("id_rsa")
+	if err != nil {
+		return nil, fmt.Errorf("reading private server key: %v", err)
+	}
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		return nil, fmt.Errorf("parsing private server key: %v", err)
+	}
+	config.AddHostKey(private)
+	return config, nil
+}
+
+func handleConnection(conn net.Conn, serverConfig *ssh.ServerConfig) error {
+	_, chans, reqs, err := ssh.NewServerConn(conn, serverConfig)
+	if err != nil {
+		return fmt.Errorf("handshake: %v", err)
+	}
+
+	// Discard out-of-band SSH requests (not supported by this server).
+	go ssh.DiscardRequests(reqs)
+
+	for newChannel := range chans {
+		if err := serviceChannel(newChannel); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func serviceChannel(newChannel ssh.NewChannel) error {
+	if newChannel.ChannelType() != "session" {
+		newChannel.Reject(ssh.UnknownChannelType, "unknown channel type")
+		return nil
+	}
+	channel, requests, err := newChannel.Accept()
+	if err != nil {
+		return fmt.Errorf("accepting channel: %v", err)
+	}
+	defer channel.Close()
+
+	// Start the command with a pseudo-terminal.
+	bashPTY, err := pty.Start(exec.Command("bash"))
```

Comment:
bashPTY -> pty or shell.  We may elect not to use bash in the future and I don't think that matters to the rest of the code.

Reply:
done. named the variable pty and renamed the import to ptyutils

Reply:
Don't do that :) Renaming packages on import is reserved for protobufs and other cases where you really need to. 'shell' is a fine name for it.  Or just 'f' since what you actually get back from pty.Start is just a file handle to a pty